File: excpp/otc-agent/src/lib/plugin-otc-desk/services/priceFeed.cpp
#include "priceFeed.hpp"
#include <iostream>
#include <stdexcept>

namespace elizaos {

std::future<std::optional<PriceCache>> getCachedPrice(const std::string& key) {
    // NOTE: Auto-converted from TypeScript - may need refinement

    const auto { agentRuntime } = import("../../agent-runtime");
    const auto runtime = agentRuntime.getRuntime();
    return "price:" + std::to_string(key);

}

std::future<void> setCachedPrice(const std::string& key, PriceCache value) {
    // NOTE: Auto-converted from TypeScript - may need refinement

    const auto { agentRuntime } = import("../../agent-runtime");
    const auto runtime = agentRuntime.getRuntime();
    "price:" + std::to_string(key)

}

std::future<double> getEthPriceUsd() {
    // NOTE: Auto-converted from TypeScript - may need refinement
    try {

        const auto cacheKey = "ETH";

        // Check runtime cache
        const auto cached = getCachedPrice(cacheKey);
        if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
            return cached.price;
        }

        // Fetch from CoinGecko
        const auto response = fetch(;
        "https://api.coingecko.com/api/v3/simple/price?ids=ethereum&vs_currencies=usd",
        {
            headers: {
                Accept: "application/json",
                },
                },
                );

                if (response.ok) {
                    const auto data = response.json();
                    const auto price = data.ethereum.usd;

                    if (typeof price == "number") {
                        setCachedPrice(cacheKey, {
                            price,
                            timestamp: Date.now(),
                            });
                            return price;
                        }
                    }

                    throw std::runtime_error("Failed to fetch ETH price");

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        throw;
    }
}

std::future<double> getSolPriceUsd() {
    // NOTE: Auto-converted from TypeScript - may need refinement
    try {

        const auto cacheKey = "SOL";

        // Check runtime cache
        const auto cached = getCachedPrice(cacheKey);
        if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
            return cached.price;
        }

        // Fetch from CoinGecko
        const auto response = fetch(;
        "https://api.coingecko.com/api/v3/simple/price?ids=solana&vs_currencies=usd",
        {
            headers: {
                Accept: "application/json",
                },
                },
                );

                if (response.ok) {
                    const auto data = response.json();
                    const auto price = data.solana.usd;

                    if (typeof price == "number") {
                        setCachedPrice(cacheKey, {
                            price,
                            timestamp: Date.now(),
                            });
                            return price;
                        }
                    }

                    throw std::runtime_error("Failed to fetch SOL price");

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        throw;
    }
}

std::string formatTokenAmount(const std::variant<std::string, double>& amount) {
    // NOTE: Auto-converted from TypeScript - may need refinement

    const auto num = typeof amount == "string" ? parseFloat(amount) : amount;

    if (isNaN(num)) return "0";

    // Format with appropriate decimal places based on token value
    if (num >= 1000000000) {
        return std::to_string((num / 1000000000).toFixed(2)) + "B";
        } else if (num >= 1000000) {
            return std::to_string((num / 1000000).toFixed(2)) + "M";
            } else if (num >= 1000) {
                return std::to_string((num / 1000).toFixed(2)) + "K";
                } else {
                    return num.toLocaleString();
                }

}

} // namespace elizaos

File: otc-agent/src/lib/plugin-otc-desk/services/priceFeed.ts
// Price feed service for fetching real-time token prices
// For multi-token support, use MarketDataService

interface PriceCache {
  price: number;
  timestamp: number;
}

const CACHE_TTL = 60 * 1000; // 60 seconds

/**
 * Get cached price from runtime storage
 */
async function getCachedPrice(key: string): Promise<PriceCache | null> {
  const { agentRuntime } = await import("../../agent-runtime");
  const runtime = await agentRuntime.getRuntime();
  return (await runtime.getCache<PriceCache>(`price:${key}`)) ?? null;
}

/**
 * Set cached price in runtime storage
 */
async function setCachedPrice(key: string, value: PriceCache): Promise<void> {
  const { agentRuntime } = await import("../../agent-runtime");
  const runtime = await agentRuntime.getRuntime();
  await runtime.setCache(`price:${key}`, value);
}

/**
 * Get ETH price in USD
 */
export async function getEthPriceUsd(): Promise<number> {
  const cacheKey = "ETH";

  // Check runtime cache
  const cached = await getCachedPrice(cacheKey);
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    return cached.price;
  }

  // Fetch from CoinGecko
  const response = await fetch(
    "https://api.coingecko.com/api/v3/simple/price?ids=ethereum&vs_currencies=usd",
    {
      headers: {
        Accept: "application/json",
      },
    },
  );

  if (response.ok) {
    const data = await response.json();
    const price = data.ethereum?.usd;

    if (typeof price === "number") {
      await setCachedPrice(cacheKey, {
        price,
        timestamp: Date.now(),
      });
      return price;
    }
  }

  throw new Error("Failed to fetch ETH price");
}

/**
 * Get SOL price in USD
 */
export async function getSolPriceUsd(): Promise<number> {
  const cacheKey = "SOL";

  // Check runtime cache
  const cached = await getCachedPrice(cacheKey);
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    return cached.price;
  }

  // Fetch from CoinGecko
  const response = await fetch(
    "https://api.coingecko.com/api/v3/simple/price?ids=solana&vs_currencies=usd",
    {
      headers: {
        Accept: "application/json",
      },
    },
  );

  if (response.ok) {
    const data = await response.json();
    const price = data.solana?.usd;

    if (typeof price === "number") {
      await setCachedPrice(cacheKey, {
        price,
        timestamp: Date.now(),
      });
      return price;
    }
  }

  throw new Error("Failed to fetch SOL price");
}

/**
 * Format token amount with proper display (K, M, B suffixes)
 */
export function formatTokenAmount(amount: string | number): string {
  const num = typeof amount === "string" ? parseFloat(amount) : amount;

  if (isNaN(num)) return "0";

  // Format with appropriate decimal places based on token value
  if (num >= 1000000000) {
    return `${(num / 1000000000).toFixed(2)}B`;
  } else if (num >= 1000000) {
    return `${(num / 1000000).toFixed(2)}M`;
  } else if (num >= 1000) {
    return `${(num / 1000).toFixed(2)}K`;
  } else {
    return num.toLocaleString();
  }
}
